src/discord/discord.py

```diff
@@ -0,0 +1,277 @@
+    result = await _discord_request(url, tok)
+
+    result_json = result.json()
+    channel = [channel for channel in result_json if channel['id'] == cid][0]
+    channel = Channel(channel['id'], channel['type'], channel['guild_id'], channel['name'], channel['permission_overwrites'])
+
+    return channel
+
+async def get_all_channels(
+    id: str = guild_id
+) -> list[str]:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/channels'
+    result = await _discord_request(url, tok)
+
+    result_json = result.json()
+    channels = [channel for channel in result_json if channel['type'] != DISCORD_CATEGORY_ID]
+    channel_names = [channel['name'] for channel in channels]
+
+    return channel_names
+
+
+async def get_role_name_by_id(
+    rid: str,
+    id: str = guild_id
+) -> str:
+    roles = await get_all_roles(id)
+    return roles[rid][0]
+
+async def get_role_by_id(
+    rid: str,
+    id: str = guild_id
+) -> dict:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/roles'
+    result = await _discord_request(url, tok)
+
+    json_s = result.json()
+    return [role for role in json_s if role['id'] == rid][0]
+
+async def get_user_roles(
+    uid: str,
+    id: str = guild_id
+) -> list[str]:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/members/{uid}'
+    result = await _discord_request(url, tok)
+
+    json_s = result.json()
+    return json_s['roles']
+
+async def get_all_roles(
+    id: str = guild_id
+) ->  dict[str, list[str]]:
+    tok = os.environ.get('TOKEN')
+    url = f'https://discord.com/api/v10/guilds/{id}/roles'
+    result = await _discord_request(url, tok)
+
+    json_s = result.json()
+    roles = [([role['id'], [role['name'], role['permissions']]]) for role in json_s]
+    return dict(roles)
+
+async def get_guild_members_with_role(
+    rid: str,
+    id: str = guild_id
+) -> list[GuildMember]:
+    tok = os.environ.get('TOKEN')
+    # base case
+    url = f'https://discord.com/api/v10/guilds/{id}/members?limit=1000'
+    result = await _discord_request(url, tok)
+
+    json_s = result.json()
+
+    matched = [GuildMember(User(user['user']['id'], user['user']['username'], user['user']['discriminator'], user['user']['global_name'], user['user']['avatar']), user['roles'])
+                    for user in json_s if rid in user['roles']] 
+
+    last_uid = matched[-1].user.id
+
+    while True:
+        url = f'https://discord.com/api/v10/guilds/{id}/members?limit=1000&after={last_uid}'
+        result = await _discord_request(url, tok)
+
+        json_s = result.json()
+        res = [GuildMember(User(user['user']['id'], user['user']['username'], user['user']['discriminator'], user['user']['global_name'], user['user']['avatar']), user['roles'])
+                    for user in json_s if rid in user['roles']] 
+        matched += res
+
+        if res == []:
```

Comment:
can we move this check before `res = ...` & make it `if len(json_s) == 0`, to be more explicit? (slightly more performant, but that's not super important)

Reply:
Handled in adeb09b55a4380b2458ddf0132ef924c1d20b612 and c8497bc28bc6b65b07b27dc5a05668839be42538.